fms/utils/serialization.py

```diff
@@ -56,6 +66,97 @@ def list_sources(architecture: str):
     return list(__adapters[architecture].keys())
 
 
+def _legacy_attn_unfused_to_fused_weight_conversion(
+    name: str, orig_sd: Mapping
+) -> Optional[Tuple[str, torch.Tensor]]:
+    """
+    function which converts unfused fms weights to fused fms weights in the case the model was using the older unfused
+    weights (version 0.0.5)
+
+    Args:
+        name: str
+            current name to convert
+        orig_sd: Mapping
+            a mapping from a name to a param, in most cases a grouping of all names to params in a module.
+
+    Returns:
+    Optional[Tuple[str, torch.Tensor]]
+        if query/key/value all exist in the given state dict, a tuple of the new fused name as well as weights will be
+        returned from this function
+        otherwise this function will return None signifying that no preprocessing needed to be done for this name param
+    """
+    # if we find query/key/value, this means the weights are unfused and therefore need to be fused
+    if "attn.query" in name or "attn.key" in name or "attn.value" in name:
+        weight_type = name.split(".")[-1]
+
+        unfused_weights = [
+            re.sub(
+                rf"attn.(query|key|value).{weight_type}",
+                f"attn.query.{weight_type}",
+                name,
+            ),
+            re.sub(
+                rf"attn.(query|key|value).{weight_type}",
+                f"attn.key.{weight_type}",
+                name,
+            ),
+            re.sub(
+                rf"attn.(query|key|value).{weight_type}",
+                f"attn.value.{weight_type}",
+                name,
+            ),
+        ]
+
+        result = (
+            re.sub(
+                rf"attn.(query|key|value).{weight_type}",
+                f"attn.in_proj.qkv_fused.{weight_type}",
+                name,
+            ),
+            torch.cat([orig_sd[w] for w in unfused_weights], dim=0),
+        )
+
+        return result
+    return None
+
+
+def simple_mapping_adapter(
+    mappers: List[Callable[[str, Mapping], Optional[Tuple[str, torch.Tensor]]]],
```

Comment:
I'm only seeing call-sites with single-entry lists of mappers. Do we actually need this function?

if so, maybe we should have a unit test that calls it with multiple mappers, at least as an example of how it works.

Reply:
Yes, we use this function in the following PR with fused MLP, we pass both the qkv fused and mlp fused to this.

Reply:
This list of mappers is only for a single parameter in the loop over all parameters (as opposed to when you nest adapters, they are running through the list of parameters multiple times.). If you were to write an adapter by hand, you would essentially run each of these mappers sequentially inside the loop. The `simple_mapping_adapter` is just hiding the loop as this case happens very often, figured it could be moved as a separate utility function

As an example: 
https://github.com/foundation-model-stack/foundation-model-stack/blob/c4d154cd0dec3d336b67a59cecdee1a23247785e/fms/models/gpt_bigcode.py#L431

https://github.com/foundation-model-stack/foundation-model-stack/blob/c4d154cd0dec3d336b67a59cecdee1a23247785e/fms/models/gpt_bigcode.py#L447